examples/multisig_plain/lib.rs

```diff
@@ -0,0 +1,837 @@
+//! through return types. Any error or invariant violation triggers a panic and therefore
+//! rolls back the transaction.
+//!
+//! ## Interface
+//!
+//! The interface is modelled after the popular gnosis multisig wallet. However, there
+//! are subtle variations from the interface. For example the `confirm_transaction`
+//! will never trigger the execution of a `Transaction` even if the treshold is reached.
+//! A call of `execute_transaction` is always required. This can be called by anyone.
+//!
+//! All the messages that are declared as only callable by the wallet must go through
+//! the usual submit, confirm, execute cycle as any other transaction that should be
+//! called by the wallet. For example to add an owner you would submit a transaction
+//! that calls the wallets own `add_owner` message through `submit_transaction`.
+//!
+//! ### Owner Management
+//!
+//! The messages `add_owner`, `remove_owner`, and `replace_owner` can be used to manage
+//! the owner set after instantiation.
+//!
+//! ### Changing the Requirement
+//!
+//! `change_requirement` can be used to tighten or relax the `requirement` of how many
+//! owner signatures are needed to execute a `Transaction`.
+//!
+//! ### Transaction Management
+//!
+//! `submit_transaction`, `cancel_transaction`, `confirm_transaction`,
+//! `revoke_confirmation` and `execute_transaction` are the bread and butter messages
+//! of this contract. Use them to dispatch arbitrary messages to other contracts
+//! with the wallet as a sender.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+use ink_lang as ink;
+
+#[ink::contract(version = "0.1.0", env = ink_core::env::DefaultEnvTypes)]
+mod multisig_plain {
+    use ink_core::{
+        env,
+        storage,
+    };
+    use ink_prelude::vec::Vec;
+    use scale::Output;
+
+    /// Tune this to your liking but be wary that many owners will not perform well.
+    const MAX_OWNERS: u32 = 50;
+
+    type TransactionId = u32;
+    const WRONG_TRANSACTION_ID: &str =
+        "The user specified an invalid transaction id. Abort.";
+
+    /// A wrapper that allows us to pass untyped parameters as blob to a `CallBuilder`
+    struct CallInput<'a>(&'a [u8]);
+
+    impl<'a> scale::Encode for CallInput<'a> {
+        fn encode_to<T: Output>(&self, dest: &mut T) {
+            dest.write(self.0);
+        }
+    }
+
+    /// A Transaction is what every `owner` can submit for confirmation by other owners.
+    /// If enough owners agree it will be executed by the contract.
+    #[derive(scale::Encode, scale::Decode, storage::Flush)]
+    #[cfg_attr(feature = "ink-generate-abi", derive(type_metadata::Metadata))]
+    #[cfg_attr(feature = "std", derive(Debug, PartialEq, Eq))]
+    pub struct Transaction {
+        /// The AccountId of the contract that is called in this transaction.
+        callee: AccountId,
+        /// The raw selector which is the function name of the `callee`that is called.
```

Comment:
I think for an learner primitive types are certainly easier to grasp because there is no need to look them up. This was my motivation to choose them here.